fsm.go

```diff
@@ -31,6 +31,24 @@ type FSM interface {
 	Restore(io.ReadCloser) error
 }
 
+// BatchingFSM extends the FSM interface to add an ApplyBatch function. This can
+// optionally be implemented by clients to enable multiple logs to be applied to
+// the FSM in batches. Up to MaxAppendEntries could be sent in a batch.
+type BatchingFSM interface {
+	// ApplyBatch is invoked once a batch of log entries has been committed and
+	// are ready to be applied to the FSM. ApplyBatch will take in an array of
+	// log entries. These log entries could be of a few types clients should
+	// check the log type prior to attempting to decode the data attached.
+	// Presently the LogCommand and LogConfiguration types will be sent.
+	//
+	// The returned slice must be the same length as the input. The returned
+	// values will be made available in the ApplyFuture returned by Raft.Apply
+	// method if that method was called on the same Raft node as the FSM.
```

Comment:
Should we suggest guidance on partial/total error? e.g. if the whole FSM is unavailable do all results need to contain the same error or just the first and the rest nil?

Reply:
All results should return the same error. I'll add this to the comment